File: antd-dart/lib/src/discover.dart
import 'dart:io';

const _portFileName = 'daemon.port';
const _dataDirName = 'ant';

/// Reads the daemon.port file written by antd on startup and returns the
/// REST base URL (e.g. "http://127.0.0.1:8082").
/// Returns empty string if the port file is not found or unreadable.
String discoverDaemonUrl() {
  final ports = _readPortFile();
  if (ports.$1 == 0) {
    return '';
  }
  return 'http://127.0.0.1:${ports.$1}';
}

/// Reads the daemon.port file written by antd on startup and returns the
/// gRPC target (e.g. "127.0.0.1:50051").
/// Returns empty string if the port file is not found or has no gRPC line.
String discoverGrpcTarget() {
  final ports = _readPortFile();
  if (ports.$2 == 0) {
    return '';
  }
  return '127.0.0.1:${ports.$2}';
}

/// Reads the daemon.port file and returns (restPort, grpcPort).
/// The file format is up to three lines:
///   line 1: REST port
///   line 2: gRPC port
///   line 3: PID of the antd process
/// A single-line file is valid (gRPC port will be 0, no PID check).
/// If a PID is present and the process is not alive, the port file is
/// considered stale and (0, 0) is returned.
(int, int) _readPortFile() {
  final dir = _dataDir();
  if (dir.isEmpty) {
    return (0, 0);
  }

  final file = File('$dir${Platform.pathSeparator}$_portFileName');
  String data;
  try {
    data = file.readAsStringSync();
  } catch (_) {
    return (0, 0);
  }

  final lines = data.trim().split('\n');
  if (lines.isEmpty) {
    return (0, 0);
  }

  // If a PID is recorded on line 3, verify the process is still alive.
  if (lines.length >= 3) {
    final pid = int.tryParse(lines[2].trim());
    if (pid != null && pid > 0 && !_isProcessAlive(pid)) {
      return (0, 0);
    }
  }

  final rest = _parsePort(lines[0]);
  final grpc = lines.length >= 2 ? _parsePort(lines[1]) : 0;
  return (rest, grpc);
}

/// Returns true if a process with the given [pid] is currently running.
///
/// On non-Windows platforms, uses `kill -0 <pid>` which sends no signal but
/// returns exit code 0 if the process exists.
/// On Windows, Dart lacks a clean way to probe a PID without side effects,
/// so we optimistically return true (trust the port file).
bool _isProcessAlive(int pid) {
  if (Platform.isWindows) {
    // No reliable non-destructive PID probe in Dart on Windows.
    return true;
  }
  try {
    final result = Process.runSync('kill', ['-0', pid.toString()]);
    return result.exitCode == 0;
  } catch (_) {
    // If we can't run the check, assume alive to avoid false negatives.
    return true;
  }
}

int _parsePort(String s) {
  final n = int.tryParse(s.trim());
  if (n == null || n < 1 || n > 65535) {
    return 0;
  }
  return n;
}

/// Returns the platform-specific data directory for ant.
///   - Windows: %APPDATA%\ant
///   - macOS:   ~/Library/Application Support/ant
///   - Linux:   $XDG_DATA_HOME/ant or ~/.local/share/ant
String _dataDir() {
  if (Platform.isWindows) {
    final appdata = Platform.environment['APPDATA'] ?? '';
    if (appdata.isEmpty) return '';
    return '$appdata${Platform.pathSeparator}$_dataDirName';
  }

  if (Platform.isMacOS) {
    final home = Platform.environment['HOME'] ?? '';
    if (home.isEmpty) return '';
    return '$home/Library/Application Support/$_dataDirName';
  }

  // Linux and others
  final xdg = Platform.environment['XDG_DATA_HOME'] ?? '';
  if (xdg.isNotEmpty) {
    return '$xdg/$_dataDirName';
  }
  final home = Platform.environment['HOME'] ?? '';
  if (home.isEmpty) return '';
  return '$home/.local/share/$_dataDirName';
}

File: antd-mcp/src/antd_mcp/discover.py
"""Port discovery for the antd daemon.

The antd daemon writes a ``daemon.port`` file on startup containing two lines:
  - Line 1: REST port
  - Line 2: gRPC port

This module reads that file using platform-specific data directory paths to
auto-discover the daemon without requiring manual configuration.
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

_PORT_FILE_NAME = "daemon.port"
_DATA_DIR_NAME = "ant"


def _data_dir() -> Path | None:
    """Return the platform-specific data directory for ant, or None."""
    if sys.platform == "win32":
        appdata = os.environ.get("APPDATA")
        if not appdata:
            return None
        return Path(appdata) / _DATA_DIR_NAME

    if sys.platform == "darwin":
        home = os.environ.get("HOME")
        if not home:
            return None
        return Path(home) / "Library" / "Application Support" / _DATA_DIR_NAME

    # Linux and other Unix-likes
    xdg = os.environ.get("XDG_DATA_HOME")
    if xdg:
        return Path(xdg) / _DATA_DIR_NAME
    home = os.environ.get("HOME")
    if not home:
        return None
    return Path(home) / ".local" / "share" / _DATA_DIR_NAME


def _read_port_file() -> tuple[int, int]:
    """Read the daemon.port file and return (rest_port, grpc_port).

    Returns (0, 0) if the file is missing or unreadable.
    A single-line file is valid; grpc_port will be 0 in that case.
    """
    data_dir = _data_dir()
    if data_dir is None:
        return 0, 0

    port_file = data_dir / _PORT_FILE_NAME
    try:
        text = port_file.read_text().strip()
    except (OSError, ValueError):
        return 0, 0

    lines = text.splitlines()
    if not lines:
        return 0, 0

    rest_port = _parse_port(lines[0])
    grpc_port = _parse_port(lines[1]) if len(lines) >= 2 else 0
    return rest_port, grpc_port


def _parse_port(s: str) -> int:
    """Parse a port string, returning 0 on failure."""
    try:
        n = int(s.strip())
        if 1 <= n <= 65535:
            return n
    except ValueError:
        pass
    return 0


def discover_daemon_url() -> str:
    """Read the daemon.port file and return the REST base URL.

    Returns ``"http://127.0.0.1:{port}"`` on success, or ``""`` if the port
    file is not found or unreadable.
    """
    rest, _ = _read_port_file()
    if rest == 0:
        return ""
    return f"http://127.0.0.1:{rest}"


def discover_grpc_target() -> str:
    """Read the daemon.port file and return the gRPC target.

    Returns ``"127.0.0.1:{port}"`` on success, or ``""`` if the port file
    is not found or has no gRPC line.
    """
    _, grpc = _read_port_file()
    if grpc == 0:
        return ""
    return f"127.0.0.1:{grpc}"
